axelor-stock/src/main/java/com/axelor/apps/stock/service/StoredProductsServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+import com.axelor.apps.stock.db.repo.MassStockMoveRepository;
+import com.axelor.apps.stock.db.repo.StockMoveLineRepository;
+import com.axelor.apps.stock.db.repo.StockMoveRepository;
+import com.axelor.apps.stock.db.repo.StoredProductsRepository;
+import com.axelor.apps.stock.exception.StockExceptionMessage;
+import com.axelor.i18n.I18n;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+
+public class StoredProductsServiceImpl implements StoredProductsService {
+
+  protected StoredProductsRepository storeProductsRepository;
+  protected MassStockMoveRepository massStockMoveRepository;
+  protected StockMoveRepository stockMoveRepository;
+  protected StockMoveLineRepository stockMoveLineRepository;
+  protected StockMoveService stockMoveService;
+  protected StockMoveLineService stockMoveLineService;
+
+  @Inject
+  public StoredProductsServiceImpl(
+      StoredProductsRepository storeProductsRepository,
+      MassStockMoveRepository massStockMoveRepository,
+      StockMoveRepository stockMoveRepository,
+      StockMoveLineRepository stockMoveLineRepository,
+      StockMoveService stockMoveService,
+      StockMoveLineService stockMoveLineService) {
+    this.storeProductsRepository = storeProductsRepository;
+    this.massStockMoveRepository = massStockMoveRepository;
+    this.stockMoveRepository = stockMoveRepository;
+    this.stockMoveLineRepository = stockMoveLineRepository;
+    this.stockMoveService = stockMoveService;
+    this.stockMoveLineService = stockMoveLineService;
+  }
+
+  @Override
+  public StoredProducts createStoredProduct(
+      Product product,
+      TrackingNumber trackingNumber,
+      BigDecimal currentQty,
+      StockLocation toStockLocation,
+      BigDecimal storedQty,
+      StockMoveLine stockMoveLine,
+      MassStockMove massStockMove) {
+    StoredProducts storedProducts = new StoredProducts();
+    storedProducts.setStoredProduct(product);
+    storedProducts.setTrackingNumber(trackingNumber);
+    storedProducts.setCurrentQty(currentQty);
+    storedProducts.setUnit(product.getUnit());
+    storedProducts.setToStockLocation(toStockLocation);
+    storedProducts.setStoredQty(storedQty);
+    storedProducts.setStockMoveLine(stockMoveLine);
+    storedProducts.setMassStockMove(massStockMove);
+    return storedProducts;
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void createStockMoveAndStockMoveLine(StoredProducts storedProducts)
+      throws AxelorException {
+    BigDecimal storedQty =
+        storedProducts.getStoredQty() != null ? storedProducts.getStoredQty() : null;
+    BigDecimal currentQty =
+        storedProducts.getCurrentQty() != null ? storedProducts.getCurrentQty() : null;
+
+    if (storedQty.compareTo(currentQty) == -1 && storedQty.compareTo(BigDecimal.ZERO) != 0) {
+      Product duplicateProduct = storedProducts.getStoredProduct();
```

Comment:
Please refactor this 
    ` Product duplicateProduct = storedProducts.getStoredProduct();
      TrackingNumber duplicateTrackingNumber = storedProducts.getTrackingNumber();
      StockLocation duplicateToStockLocation = storedProducts.getToStockLocation();
      StockMoveLine duplicateStockMoveLine = storedProducts.getStockMoveLine();
      MassStockMove duplicateMassStockMove = storedProducts.getMassStockMove();
      storedProducts.setCurrentQty(storedQty);
      StoredProducts duplicateStoredProduct =
          createStoredProduct(
              duplicateProduct,
              duplicateTrackingNumber,
              currentQty.subtract(storedQty),
              duplicateToStockLocation,
              BigDecimal.ZERO,
              duplicateStockMoveLine,
              duplicateMassStockMove);
      storeProductsRepository.save(duplicateStoredProduct);` in a method with a explicit name